performance_manager/lib/l1_full_trip_events.py

```diff
@@ -0,0 +1,223 @@
+        )
+
+        if event_type.merge_dataframe.shape[0] == 0:
+            event_type.merge_dataframe = pandas.DataFrame(
+                columns=["pk_id", "hash"]
+            )
+        else:
+            event_type.merge_dataframe = add_event_hash_column(
+                event_type.merge_dataframe,
+                expected_hash_columns=expected_hash_columns,
+            )[["pk_id", "hash"]]
+
+        event_type.merge_dataframe.rename(
+            columns={"pk_id": event_type.fk_rename_field}, inplace=True
+        )
+
+
+def merge_events(events_list: Dict[str, EventsToMerge]) -> pandas.DataFrame:
+    """
+    merge all unprocessed event types into single dataframe
+    """
+    return_dataframe = events_list["moving_vp"].merge_dataframe.merge(
+        events_list["stopped_vp"].merge_dataframe, how="outer"
+    )
+    return_dataframe = return_dataframe.merge(
+        events_list["stopped_tu"].merge_dataframe, how="left"
+    )
+
+    for column in return_dataframe.columns:
+        return_dataframe[column] = return_dataframe[column].astype("Int64")
+
+    return_dataframe = return_dataframe.fillna(numpy.nan).replace(
+        [numpy.nan], [None]
+    )
+
+    return return_dataframe
+
+
+def insert_and_update_database(
+    insert_dataframe: pandas.DataFrame, db_manager: DatabaseManager
+) -> None:
+    """
+    update / insert full trip event rds L1 table, handled as delete and insert
+    """
+    if insert_dataframe.shape[0] == 0:
+        logging.warning("No new Full Trip Event records to load")
+        return
+
+    db_manager.truncate_table(TempFullTripEvents)
+
+    db_manager.insert_dataframe(insert_dataframe, TempFullTripEvents.__table__)
+
+    delete_query = sa.delete(FullTripEvents.__table__).where(
+        FullTripEvents.hash.in_(sa.select(TempFullTripEvents.hash))
+    )
+    db_manager.execute(delete_query)
+
+    insert_cols = [
+        "hash",
+        "fk_vp_stopped_event",
+        "fk_vp_moving_event",
+        "fk_tu_stopped_event",
+    ]
+
+    insert_query = (
+        FullTripEvents.metadata.tables[FullTripEvents.__tablename__]
+        .insert()
+        .from_select(
+            insert_cols,
+            sa.select(
+                TempFullTripEvents.hash,
+                TempFullTripEvents.fk_vp_stopped_event,
+                TempFullTripEvents.fk_vp_moving_event,
+                TempFullTripEvents.fk_tu_stopped_event,
+            ).where(
+                (TempFullTripEvents.hash.notin_(sa.select(FullTripEvents.hash)))
+            ),
+        )
+    )
+    db_manager.execute(insert_query)
+
+
+def process_full_trip_events(db_manager: DatabaseManager) -> None:
+    """
+    process new events from L0 tables and insert to L1 full trip event RDS table
+    """
+    try:
+        events_list = get_merge_list()
+        load_and_transform(events_list, db_manager)
+        insert_dataframe = merge_events(events_list)
+        insert_and_update_database(insert_dataframe, db_manager)
```

Comment:
i think the flow in this script makes sense and i could sort the majority of it out eventually. however, i do think the whole thing could use more comments, especially to help out future us.

at least for me, what the pandas and sqa alchemy code isn't quickly decipherable, so a larger comment on the `process_full_trip_events` with the overall strategy of this script and inline comments on the other functions would be helpful.